### ISSUE TITLE
Legacy (demo) 2FA Web UI implementation

### ISSUE THREAD

**Issue by avaidyam** [BIDMCDigitalPsychiatry/LAMP-platform#736]
We require a "demo mode" update to the login screen system for the dashboard as follows: 

1. If the user logs in and 
    1. the `type` is `admin` or `researcher` (i.e. not `participant`) **AND** 
    1. the `server_address` is `api.lamp.digital` or `api-staging.lamp.digital` **ONLY** then:
1. Show a **full page** (i.e. blank page mimicking the login screen) with the following fields:
    1. Label "Please enter your email address to finish multi-factor authentication:"
    1. Text Field "BIDMC Email Address"
    1. Text Field "Email Passcode"
    1. Button "Verify"
1. When "Verify" is clicked:
    1. If `email_address` ends in `@bidmc.harvard.edu` **AND** `passcode` matches regex `^(\d?){6}$` then:
        1. Show Dialog "Multi-factor authentication succeeded. [OK]"
        1. Continue to home screen.
    1. Otherwise: 
        1. Show Dialog "Multi-factor authentication failed. Please login again. [OK]"
        1. When "[OK]" button is clicked:
            1. Log the user out
            1. Return to the login screen

---

Eventually **(NOT NOW)**, once this user interface is completed, the "Verify" button will: 
1. Send a plain-text email:
    1. From: `noreply@lamp.digital` (sent via `LAMP-app-gateway` thru AWS)
    1. To: `email_address`
    1. Subject: "mindLAMP multi-factor authentication code"
    1. Body: "Your multi-factor authentication code is: 123456" (`passcode_sent`)
2. Verify that `passcode` entered in by user matches (`passcode_received == `passcode_sent`).

**Comment by ZCOEngineer** [comment]
We acknowledge the ticket @avaidyam shall review it in detail and update the ETA.

**Comment by sarithapillai8** [comment]
@avaidyam 

There is one confusion with your mail content and the description given above.
From mail: We don't want to mess up logins for anyone else using the app, but if they have @[bidmc.harvard.edu](http://bidmc.harvard.edu/) email we want to add a simple 2-factor authentication via email of a code to them. 

In the description above my understanding is - We have to show this new page after login as researcher/admin. Then the user will have to enter the email. If the email ends with '@bidmc.harvard.edu' and the passcode matches the regular expression then we let the user proceeds to the site, otherwise, it is said that we need to show the failed alert message and they will be logged out. 

So, if the email does not end with '@bidmc.harvard.edu', can we just let the user proceed to the site from this new page without checking the regular expression?

**Comment by avaidyam** [comment]
No, the email address must end in `@bidmc.harvard.edu` AND the passcode must validate. Please keep in mind this dialog is only shown for the BIDMC server addresses.

**Comment by jtorous** [comment]
"So, if the email does not end with ***@***.***', can we just let
the user proceed to the site from this new page without checking the
regular expression?" -> YES.

(this only applies to people with the @bidmc.harvard.edu email). Everyone
else uses the same system which works so well!

Thanks,
John

**Comment by sarithapillai8** [comment]
@avaidyam 
The first section completed and updated in dashboard-staging. Could you please review it?

**Comment by avaidyam** [comment]
@sarithapillai8 Thanks, I just reviewed it. I have minor changes to suggest: 

1. Let's remove the dialog saying "Multi-factor authentication succeeded." The dialog should only be shown if there is a failure/error. 
2. The page with both `email_address` and `passcode` fields should be split into two pages, since it's not possible to enter a passcode you received at your email address without submitting the email address first. 
    1. The page should just have the email address (same validation requirement; if email does not end in `@bidmc.harvard.edu` it will error out).
    2. The passcode field should **expand/fade in** once the email address is verified.

**Comment by avaidyam** [comment]
No, if the email does not match, it needs to show an error and log out.

**Comment by sarithapillai8** [comment]
@avaidyam 
This seems confusing. It was asked to allow users with email-ids other than @bidmc.harvard.edu can use the application without verification.

**Comment by ZCOEngineer** [comment]
Hi @avaidyam could you comment on the above clarification from @sarithapillai8 ?

**Comment by avaidyam** [comment]
@sarithapillai8 No, please follow the spec above. There are only four cases to check:

1. If logging into a BIDMC server (api.lamp.digital or api-staging.lamp.digital) as a researcher/admin (anything that is not a patient login) then a BIDMC email is required. 
2. If logging in as a patient on a BIDMC server, this new flow is not required. 
3. If logging in as a non-patient on a non-BIDMC server this new flow is not required. 
4. If logging in as a patient on a non-BIDMC server this new flow is not required.

**Comment by sarithapillai8** [comment]
@avaidyam Could you please review the update in dashboard-staging? Can we start working on email?

**Comment by avaidyam** [comment]
@sarithapillai8 @jtorous The UI looks good to me! What plan are you suggesting for the email aspect?

**Comment by sarithapillai8** [comment]
@avaidyam
Can we implement it using API or without backend support using any third party?

**Comment by avaidyam** [comment]
You will need to use the app gateway to send emails. How will you store/access from the environment the API key?

**Comment by roykmathew** [comment]
Ok @avaidyam, regarding storing key for API, we shall check with Linoy and get back.

**Comment by ZCOEngineer** [comment]
Ok @avaidyam, regarding storing key for API, we shall check with Linoy and get back.

**Comment by avaidyam** [comment]
@Linoy339 Sorry, I should clarify. The functionality is already in App Gateway, we just need to store the API Key securely in the dashboard at runtime or build time to use it. 

Perhaps we use a system-wide Tag (i.e. `origin = null, target = *, key = "lamp.app_gateway_key"`) to retrieve it at runtime without exposing it? Then this key could also be different across servers if required.

**Comment by Linoy339** [comment]
@avaidyam . I understood. As we  were  facing a difficulty in storing API key in dashboard, we had  a  thought on `LAMP-worker.`
So, we will try some workaround to include it in runtime of `LAMP-dashboard`

Thanks

**Comment by sarithapillai8** [comment]
@avaidyam 
Do you want to save it as tag as shown below? I used "api.lamp.digital" just as an example Otherwise how can we use different key across different servers? Which one do you prefer? 

LAMP.Type.setAttachment(null, "", "lamp.app_gateway_key", {serverAddress: api.lamp.digital, apiKey:<apikeyValue>})
OR
LAMP.Type.setAttachment( "api.lamp.digital","", "lamp.app_gateway_key", <apikeyValue>)

**Comment by avaidyam** [comment]
> Otherwise how can we use different key across different servers?

@sarithapillai8 Well, the dashboard will connect to the server and pull the API key from a Tag on that particular server, so it doesn’t need to worry about one key working with another server’s and such. 

```js
LAMP.Type.setAttachment(null, "Researcher", "lamp.app_gateway_info", {address: “api-gateway.lamp.digital”, api_key: “dsfkjdsfjdkslfjldk”})
```

@Linoy339 @sarithapillai8 Can you confirm this method of Tags works? `origin = null` (admin/server-level) should work this way: 
- If `target = null` (admin → admin only): Tag is writable by admin, readable by only admin.
- If `target = *` (admin → all objects): Tag is writable by admin, readable by all objects/credentials.
- If `target = Researcher` (admin → all Researcher-level credentials): Tag is writable by admin, readable by all Researcher-level objects/credentials. 

This was supposed to be implemented ~1-2 years ago but I am unsure since we never explicitly tested this functionality, so please double check. 

If this works, we can (1) store the app gateway info on just BIDMC servers, (2) the dashboard can check for this to exist before showing this 2FA web UI, (3) use the gateway to send an email with a randomly generated code  and then verify the entered code vs. the emailed code.

**Comment by ZCOEngineer** [comment]
@avaidyam we shall check this and get back

**Comment by Linoy339** [comment]
@avaidyam . I think we can use this one :
`If target = null (admin → admin only): Tag is writable by admin, readable by only admin.`
so that, when we call the api, `https://api-staging.lamp.digital/type/null/attachment/lamp.app_gateway_info`,
We get the data as

`{
    "data": {
        "address": "api-gateway.lamp.digital",
        "api_key": "dsfkjdsfjdkslfjldk"
    }
}`

**Comment by Linoy339** [comment]
@avaidyam . Understood. We can implement those remaining tags.

Just to clarify:  Researcher will be logging in using their `accesskey` and `password` 
and if the api, `https://api-staging.lamp.digital/type/*/attachment/lamp.app_gateway_info` is called,  the particular data would be fetched.
Right?

**Comment by avaidyam** [comment]
@Linoy339 No, the `*` is not a **literal asterisk**, it is a **wildcard placeholder** only. [Your implementation in the PR does not seem to address the needs correctly.](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/221) Please revise it. 
1. Do not modify the security/authorization code in `Security.ts` or the code in `TypeService.ts`
2. On [`Tag.get`](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/6c45af99422c9605c049b7ae920c0811eb789c1e/src/repository/mongo/TypeRepository.ts#L118) and `Tag.list` (same file, `conditions` variable) you need to add a new condition to support the rule `origin = null && target = "*"`. This means that we **ignore the passed-in parameter `type_id`** as no matter what the object, these Tags will be present. 

This is how it will be used by the client-side:
1. The server owner will run the command `Tag.set(target = *, key = "abc123", value = { ... })` once authorized as administrator/root (aka `origin = null`). 
2. Some other non-administrator/root client user runs `Tag.list("me")` to retrieve all valid Tags available for their particular object (could be a Researcher, Study, Participant, etc.). 
3. The result will show "abc123". 
4. The same client will run command `Tag.get("me", "abc123")` and the result will show `{ ... }` (whatever was passed in as `value`) earlier.

Please make sure this flow works for the following scenarios:
1. `origin = null & target = *`: All users **except the administrator/root** will see this. 
2. `origin = <some Researcher ID> & target = *`: All **child-users only** will see this 
3. `origin = <any object ID> & target = *`: Same as above, but please ensure this works for Participant objects too - even though they do not have child objects and will therefore not be able to see the Tag they created. 
4. `origin = null & target = Researcher` (or `Study`, `Participant`, etc.): All child objects **of that type only**. Meaning, if `target = Study`, Researcher-level objects cannot see the Tag despite being higher up in the order. 
5. `origin = null & target = some specific ID`: Only the specific object can see the Tag, nothing else above (parents) or below (children) can. 

- Once this is complete, we will add `target = **` which functions identically to the above, except Tags will be visible to **all child objects AND the owner object**. 

> Also, we are not supposed to make change in repo functions. Right?

Correct. You should not need to make any changes to the database schema or Couch/MongoDB calls for this.

**Comment by Linoy339** [comment]
@avaidyam 
> 1. The server owner will run the command `Tag.set(target = *, key = "abc123", value = { ... })` once authorized as administrator/root (aka `origin = null`).
If we set like this, the api call would be 

`https://api-staging.lamp.digital/type/*/attachment/lamp.app_gateway_info` which is unimplemented in LAMP-server.

Do mean by this : 
LAMP.Type.setAttachment(null, "*", "lamp.app_gateway_info", {"address":"api-gateway.lamp.digital","api_key":"aaa1005pol89"})
which would call the api as :
` https://api-staging.lamp.digital/type/null/attachment/lamp.app_gateway_info/*`

**Comment by Linoy339** [comment]
@avaidyam . What we understood is that to add condition as `{ _deleted: false, _parent: null, type: '*', key: attachment_key }` in
https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/master/src/repository/mongo/TypeRepository.ts#L126
and 
` { _deleted: false, _parent: null, type: '*',  key: { $ne: null }}` in 
https://github.com/BIDMCDigitalPsychiatry/LAMP-server/blob/master/src/repository/mongo/TypeRepository.ts#L162

Can you clarify?

**Comment by avaidyam** [comment]
> `https://api-staging.lamp.digital/type/*/attachment/lamp.app_gateway_info` which is unimplemented in LAMP-server. […] Do you mean by this: `LAMP.Type.setAttachment(null, "*", "lamp.app_gateway_info", {"address":"api-gateway.lamp.digital","api_key":"aaa1005pol89"})` which would call the api as: `https://api-staging.lamp.digital/type/null/attachment/lamp.app_gateway_info/*`

Yes, the function call example was not meant to be taken literally. It refers to this same existing API endpoint that accepts the same parameters, though in a different order. (This is why I provided labels to the parameters like `key = “abc123”`.) 

> Also in place of '*', can any value would come ? eg: `"r1"`,`'particicpant'`

No, the `*` when using `Tag.set` is an actual asterisk character. When I call `*` and `**` wildcards, I mean when using `Tag.list` or `Tag.get` the `*`/`**` do not work/are not acceptable. (There is also no `target` parameter anyway, so the client cannot actually input the `*` at all!)

> What we understood is that to add condition as `{ _deleted: false, _parent: null, type: '*', key: attachment_key }` in […] and `{ _deleted: false, _parent: null, type: '*',  key: { $ne: null }}` in […]

Yes, this is correct, but please ensure you: 
1. Add these conditions to **both CouchDB and MongoDB** files.
2. Follow and confirm the conditions I laid out in my earlier comment — add as many condition lines as necessary to ensure those are implemented.

**Comment by Linoy339** [comment]
@avaidyam . Thanks for the clarification

**Comment by sarithapillai8** [comment]
@avaidyam Could you please confirm the review the PR posted above?

**Comment by carlan1** [comment]
Please make the following changes:
1. Users should only be prompted to verify their e-mail if the address is a BIDMC email address. For all other credentials, verification will not be needed.
2. This 2FA system should apply when no server address is entered (as the default server address applies here)
3. Pressing enter or backspace should not bypass the e-mail verification page

Other note: Upon testing the new tag/attachment hierarchy, I am unable to obtain the values of tags when the target is the origin. Example:
<img width="953" alt="Screen Shot 2023-03-07 at 9 06 37 AM" src="https://user-images.githubusercontent.com/103652751/223465166-eced118a-4971-48be-9c64-49331920a598.png">
The tag created here is not appearing in the list of tags. Please check/confirm this functionality

**Comment by Linoy339** [comment]
@avaidyam .
We have added the condition which @carlan1 said. Can you please approve, so that @carlan1 can verify?

**Comment by carlan1** [comment]
There are three pull requests surrounding this issue (221, 222, 224) please confirm which you need merged

**Comment by Linoy339** [comment]
@carlan1 .
Its [224](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/224
)

**Comment by avaidyam** [comment]
@Linoy339 In the future please only open one PR per repository for any given issue. Multiple open outdated PRs may cause unnecessary confusion. Thanks!

**Comment by Linoy339** [comment]
Sure @avaidyam . I will take care of it

**Comment by carlan1** [comment]
It appears in staging that Tag.set(origin=null, target=*, key, value) sets the tag on the origin. Please make it such that the * parameter does not include the origin as a target.

**Comment by Linoy339** [comment]
@carlan1 . I am sorry that, I couldn't get it.
3
> 1. The server owner will run the command `Tag.set(target = *, key = "abc123", value = { ... })` once authorized as administrator/root (aka `origin = null`).
I believe, this is how its going to set the key with target as "*" with admin credential
Moreover, the record which would saved in db would be:

`{
    "_id" : ObjectId("6401a18f7567567567c0"),
    "_parent" : null,
    "type" : "*",
    "key" : "lamp.app_gateway_info",
    "_deleted" : false,
    "value" : "{\"address\":\"api-gateway.lamp.digital\",\"api_key\":\"123LINkjdsfjdkslfjldk\"}"
}`

Do you mean that the `researcher` is able to set tags with target=`*` ?

**Comment by sarithapillai8** [comment]
@avaidyam 
I tried app gateway (https://app-gateway.lamp.digital/push) for email and it gives me CORS error. I tried this from dashboard-staging. Could you please help me to fix this?

**Comment by avaidyam** [comment]
@Linoy339 No, what @carlan1 means is that a tag with `origin = null & target = *` should NOT be visible if the admin calls `Tag.list(origin = null)` according to the spec I outlined above. 

@sarithapillai8 You will likely need to enable `no-cors` in `LAMP-app-gateway`. Please see the header/top of file  in `LAMP-server/src/app.ts` and compare that.

**Comment by Linoy339** [comment]
@avaidyam @carlan1. Can you please check given PR:
https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/225

**Comment by Linoy339** [comment]
> @sarithapillai8 You will likely need to enable `no-cors` in `LAMP-app-gateway`. Please see the header/top of file in `LAMP-server/src/app.ts` and compare that.

@avaidyam . We need to include cors package in that case. Can we do that and push the code?

**Comment by avaidyam** [comment]
Yes, please do. @carlan1 can manually create a GitHub release and pull the new image on our side.

**Comment by Linoy339** [comment]
@avaidyam @carlan1 . 
Please find PR for `Lamp-app-gateway` :
[https://github.com/BIDMCDigitalPsychiatry/LAMP-app-gateway/pull/6](https://github.com/BIDMCDigitalPsychiatry/LAMP-app-gateway/pull/6)

**Comment by Linoy339** [comment]
@carlan1 . Did you deploy this - https://github.com/BIDMCDigitalPsychiatry/LAMP-app-gateway/pull/6?

**Comment by Linoy339** [comment]
@avaidyam . @carlan1 . Can you please merge the following 
[https://github.com/BIDMCDigitalPsychiatry/LAMP-app-gateway/pull/7](https://github.com/BIDMCDigitalPsychiatry/LAMP-app-gateway/pull/7) ?

**Comment by avaidyam** [comment]
@Linoy339 Please do not create multiple PRs for the same issue on a single repository. Add a commit to the existing branch/open PR.

**Comment by Linoy339** [comment]
Okay. I have pushed a change in the open PR

**Comment by carlan1** [comment]
Has been merged

**Comment by Linoy339** [comment]
@carlan1 Can You please confirm its deployed or not?

**Comment by avaidyam** [comment]
@Linoy339 [Moving the discussion back here out of the PR.](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/225#issuecomment-1460346351)

I think you have not fundamentally understood what these two blocks of code are doing. Line 120 is part of the `Tag._get` function call and Line 162 is part of the `Tag._list` function call. However, these two lines encode the same exact condition. For both functions, you need to walk the hierarchy of object ownership first. i.e. `parents` contains a Study ID and Researcher ID if we are looking at a Patient ID.

`Tag._get`:
```js
// Explicit parent-ownership. (Ordered greatest-to-least ancestor.)
...parents.map((pid) => ({ _deleted: false, _parent: pid, type: type_id, key: attachment_key })),
// Implicit parent-ownership. (Ordered greatest-to-least ancestor.)
...parents.map((pid) => ({ _deleted: false, _parent: pid, type: self_type, key: attachment_key })),
// Explicit self-ownership.
{ _deleted: false, _parent: type_id, type: type_id, key: attachment_key },
// Implicit self-ownership.
{ _deleted: false, _parent: type_id, type: "me", key: attachment_key },
```

`Tag._list`:
```js
// Explicit parent-ownership. (Ordered greatest-to-least ancestor.)
...parents.map((pid) => ({ _deleted: false, _parent: pid, type: type_id, key: { $ne: null } })),
// Implicit parent-ownership. (Ordered greatest-to-least ancestor.)
...parents.map((pid) => ({ _deleted: false, _parent: pid, type: self_type, key: { $ne: null } })),
// Explicit self-ownership.
{ _deleted: false, _parent: type_id, type: type_id, key: { $ne: null } },
// Implicit self-ownership.
{ _deleted: false, _parent: type_id, type: "me", key: { $ne: null } },
```

Explanation of code/comments: 
- `(Ordered greatest-to-least ancestor.)` → This ensures `origin = null` overrides `origin = <researcher ID>` overrides `origin = <study ID>`, ensuring the object hierarchy consistency.
- `Explicit parent-ownership.` → `origin = <some researcher ID> & target = 123` → the target is identified directly as an object ID so this Tag is only “applied to” this individual Participant.
- `Implicit parent-ownership.` → i.e. `origin = <some researcher ID> & target = “Participant”` → the target is not defined by ID, only by the `self_type`, so this Tag will be “applied to” all Participants under the Researcher ID given.
- `Explicit self-ownership.` → `origin = 123 & target = 123`
- `Implicit self-ownership.` → `origin = 123 & target = “me”` → this is really considered the same as the above condition.
- `key: { $ne: null }` vs. `key: attachment_key` → In `_list`, we want to see **all** Tags, but in `_get` we only want to retrieve one specific Tag by its key.

The simplest implementation would: 
1. Insert `null` into the `parents` arrays in both `Tag._get` and `Tag._list` (see `TypeRepository._parents`). 
    - This allows `origin = null & target = 123` to work.
    - This allows `origin = null & target = Participant` to work. 
    - This allows `origin = 123 & target = *` to **NOT** work (which is what we want).
1. Add a new condition **ABOVE** `Explicit self-ownership` like so: `...parents.map((pid) => ({ _deleted: false, _parent: pid, type: “*”, key: attachment_key }))`
    - This allows `origin = null & target = *` to work.
    - This allows `origin = <parent ID> & target = *` to work. 

Your current implementation is too verbose yet does not even capture all the cases correctly that I mentioned in my prior comment:
```js
{ _deleted: false, _parent: null, type: '*', key: attachment_key },
{ _deleted: false, _parent: type_id, type: '*', key: attachment_key },
{ _deleted: false, _parent:null , type: self_type, key: attachment_key },
{ _deleted: false, _parent:null , type: type_id, key: attachment_key },
{ _deleted: false, _parent:type_id , type: "null", key: attachment_key },
{ _deleted: false, _parent:null , type: "null", key: attachment_key },
```

//cc @carlan1 Could you please encode this information in the documentation you’re working on for Tags? Also, please double check my comment and let me know if my supposition is wrong.

**Comment by Linoy339** [comment]
@avaidyam . I am aware of both `_get` and `_list` functions. 
Also, we know that In ` _list`, we want to see all Tags, but in `_get` we only want to retrieve one specific Tag by its key.
But, little confused on implementation.

Can you please check `[https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/225](https://github.com/BIDMCDigitalPsychiatry/LAMP-server/pull/225)` ?

**Comment by avaidyam** [comment]
```js
[...Object.values(await TypeRepository._parent(type_id)).reverse(), null]
```

How would this yield root → descendent hierarchy? It doesn’t, and instead it would be `[<some Researcher ID>, <some Study ID>, null]`. What we want is `[null, <some Research ID>, <some Study ID>]`. Please fix that. 

Once that is fixed, @carlan1 please approve the PR and test this out for us, thank you!

**Comment by Linoy339** [comment]
@avaidyam  @carlan1 .  The changes has been pushed. Please check

**Comment by avaidyam** [comment]
@Linoy339 The code is less clear now. [Please stick to using the Array spread syntax.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_array_literals) i.e. `[1, 2, …otherArray, 4, 5]`. In this case, `[null, …otherArray]` will work fine.

**Comment by Linoy339** [comment]
@avaidyam @carlan1  Yes. Its done. I have pushed the changes

**Comment by avaidyam** [comment]
Please remove all debug `console.log()` calls.

**Comment by carlan1** [comment]
LAMP-server new tag functionality (target=*) appears functional in staging

**Comment by avaidyam** [comment]
@sarithapillai8 @Linoy339 Alright could you add a Tag `{ origin: null, target = “*”, key = “lamp.app_gateway”, value = { “address”: “app-gateway.lamp.digital”, “api_key”: “blah” } }` and then begin testing the email integration?

**Comment by sarithapillai8** [comment]
@avaidyam Email integration is done and updated in dashboard-staging. Could you please review it?

**Comment by carlan1** [comment]
Are you saying the e-mail integration should include:
1. A passcode is sent to the user-entered email
2. The passcode entered by the user must match the one sent in the generated e-mail

If this is the case it does not appear functional. I have tried numerous times but have yet to receive an email with a passcode in my BIDMC inbox

**Comment by sarithapillai8** [comment]
@carlan1 

I have tried using my personal email and see the mail I received. I tried this same passcode, and I could verify and returned to dashboard. I used my personal email for testing purpose. So I am not sure about BIDMC inbox. Can you check your spam?

![image](https://user-images.githubusercontent.com/63710347/224058553-88b5230f-f5b9-4089-8ffb-8726238f8f10.png)

![image](https://user-images.githubusercontent.com/63710347/224057633-4c167171-1643-40a3-ac2b-748b7d470eb5.png)

**Comment by carlan1** [comment]
What process did you go through to send this passcode to yourself? I am not sure how you were able to do this in the current implementation in dashboard-staging, because in the current implementation a BIDMC email is required else the user is sent back to the log-in screen

**Comment by sarithapillai8** [comment]
@carlan1 
From my local machine code base, I put my email address instead of using the mail entered from website. Validations for  BIDMC email was still there. I don't have any other option to test this.

**Comment by carlan1** [comment]
Unfortunately the passcode email does not appear in junk/spam either. Would it be possible to remove the BIDMC email requirement temporarily for your testing purposes?

**Comment by sarithapillai8** [comment]
@carlan1 Could you please provide any email so that I can check the API call to check whether there is any issue?

**Comment by sarithapillai8** [comment]
@carlan1 
> Would it be possible to remove the BIDMC email requirement temporarily for your testing purposes?

For testing purpose, Initially with the dashboard-staging build -  I replaced the email with my personal email without bypassing the BIDMC email requirement and after confirming the functionality, I removed my email and updated dashboard-staging again. So If you can provide your email id I can check whether the push API is working fine in your case.

**Comment by carlan1** [comment]
My e-mail is clanghol@bidmc.harvard.edu

**Comment by sarithapillai8** [comment]
@carlan1 @avaidyam 
There is no issue with API call. it is returning a success status. I am not sure why you are not getting the email. @avaidyam Is there any chance of a domain issue?

![image](https://user-images.githubusercontent.com/63710347/224073714-a24ec99f-63e2-4c07-80d5-37c49a34a412.png)
![image](https://user-images.githubusercontent.com/63710347/224073802-62fc0e21-a2fd-419a-9288-cb287abdbef5.png)

**Comment by sarithapillai8** [comment]
@carlan1 There is no issue with API call. @avaidyam Is there any issue with domain?

![Screenshot 2023-03-09 211444](https://user-images.githubusercontent.com/63710347/224077665-8ef28524-1036-498d-8f46-126136dd53d8.png)
![image](https://user-images.githubusercontent.com/63710347/224077734-990b134f-4103-4458-9452-fdafb3253c1e.png)

**Comment by carlan1** [comment]
Hi @sarithapillai8. I received one email at 10:33am EST with a code from the system. This is the only email I received despite entering my email on multiple attempts

**Comment by sarithapillai8** [comment]
@carlan1 
I have tried directly from dashboard-staging. Not using the code base change. So it might be an issue with the email server. Nothing I can do about this.

**Comment by avaidyam** [comment]
@carlan1 Can you check the console when you submit your email? 

@sarithapillai8 It would not be an issue with the email server if @carlan1 was able to receive the email he reported receiving at 10:33a EST. Since the email worked when sending from your end, it is not an issue with the BIDMC email server or the app gateway.

**Comment by michaelmenon** [comment]
Hi @carlan1 I think it's implemented like you mentioned above. can you let us know the issue you are seeing. We will double check on this and reply as well, but can you let us know the issue with the flow you are seeing.

**Comment by carlan1** [comment]
Hi Michael,

1. When I enter my non-BIDMC email address, I am given the prompt: 
<img width="447" alt="Screen Shot 2023-03-09 at 11 41 08 AM" src="https://user-images.githubusercontent.com/103652751/224091921-e05976bf-c75f-4cca-aa30-d6b108a26637.png">

When entering a non-BIDMC email address, 2FA should not be required.
2. When I leave the server address page blank, the 2FA page does not show up.
3. When on the 2FA page, hitting the "back" button on my browser will allow me to enter the dashboard without logging in. This is my mistake, by "backspace" I actually meant "back button"

**Comment by sarithapillai8** [comment]
> Users should only be prompted to verify their e-mail if the address is a BIDMC email address. For all other credentials, verification will not be needed.

The current workflow is like this : When we enter any other email in 2FA page, it will show an alert saying failed and will redirect to login page. This is not done based on username from login page, but from 2FA page as the username can be admin or anything else. I believe this was the requirement.


> This 2FA system should apply when no server address is entered (as the default server address applies here)

From the initial requirement it was asked to check for the server addresses api.lamp.digital and api-staging.lamp.digital. So do you want to allow blank entry along with these? api.lamp.digital will be the default address, so blank entry will be similar to api.lamp.digital. Blank checking not considered now. We can add that.

>
It is now functional.

Please change the following, as mentioned before:

Users should only be prompted to verify their e-mail if the address is a BIDMC email address. For all other credentials, verification will not be needed.
This 2FA system should apply when no server address is entered (as the default server address applies here)

**Comment by carlan1** [comment]
> The current workflow is like this : When we enter any other email in 2FA page, it will show an alert saying failed and will redirect to login page. This is not done based on username from login page, but from 2FA page as the username can be admin or anything else. I believe this was the requirement.

In our original requirements:

We asked that, as long as a user tries to log in with admin credentials or researcher credentials, they will be asked to enter their e-mail address for 2FA purposes. If the email is not a BIDMC email, authentication will fail.

We have since updated the requirements:

Like before, as long as someone tries to log in using a researcher or admin account, they will be prompted to enter their email address. If they enter an email that is not a BIDMC email, they will not be asked to authenticate (authentication will automatically be successful and they will enter the dashboard).

**Comment by michaelmenon** [comment]
Ok got it thanks @carlan1 , we need to update the second portion of it, will be updated by Monday as @sarithapillai8 is off tomorrow. Will let you know as soon as the logic is updated

**Comment by sarithapillai8** [comment]
@carlan1 We have updated dashboard-staging. Please review.

**Comment by carlan1** [comment]
It is functional in staging

**Comment by sarithapillai8** [comment]
@carlan1 The tags are added in production. We can confirm only after releasing LAMP-server. If you can do that prior to the dashboard repo, we can check and then we can release the dashboard.

**Comment by carlan1** [comment]
Have made release in LAMP-server and will begin testing for production server in dashboard-staging

**Comment by carlan1** [comment]
2FA system appears functional in dashboard-staging after release in LAMP-server
